lib/espresso-runner.js

```diff
@@ -66,29 +70,39 @@ class EspressoRunner {
   }
 
   async startSession (caps) {
-    let cmd = ['am', 'instrument', '-w', '-e', 'debug', process.env.ESPRESSO_JAVA_DEBUG === 'true' ? 'true' : 'false',
+    let cmd = ['shell', 'am', 'instrument', '-w', '-e', 'debug', process.env.ESPRESSO_JAVA_DEBUG === 'true' ? 'true' : 'false',
       `${TEST_APK_PKG}/android.support.test.runner.AndroidJUnitRunner`];
 
-    logger.info(`Starting Espresso Server v${version} with cmd: ${cmd.join(' ')}`);
+    logger.info(`Starting Espresso Server v${version} with cmd: adb ${cmd.join(' ')}`);
 
-    // start the instrumentation process, but do not wait for it
-    // otherwise it will block for longer than needed for the device to be ready
-    let err;
-    this.adb.shell(cmd).catch((e) => { // eslint-disable-line
-      // handle asynchronous errors that no longer get thrown
-      err = e;
+    // start the instrumentation process
+    this.instProcess = this.adb.createSubProcess(cmd);
+    this.instProcess.on('exit', (code, signal) => {
+      logger.info(`Instrumentation process exited with code ${code} from signal ${signal}`);
+    });
+    this.instProcess.on('die', (code, signal) => {
+      logger.error(`Instrumentation process died with code ${code} and signal ${signal}`);
+    });
+    this.instProcess.on('stream-line', line => {
+      logger.debug(`[Instrumentation]${line.trim()}`);
+    });
+
+    await this.instProcess.start((stdout, stderr) => {
+      if (stdout.includes('io.appium.espressoserver.EspressoServerRunnerTest:')) {
+        return true;
+      }
+      if (stderr.includes('Exception')) {
+        throw new Error(stderr.trim());
+      }
```

Comment:
What is going to happen if none of the conditions have been reached?

Reply:
@mykola-mokhnach a timeout error will be thrown